Packs/QualysVulnerabilityManagement/Integrations/QualysVulnerabilityManagement/QualysVulnerabilityManagement.py

```diff
@@ -0,0 +1,931 @@
+
+    return None, readable_output
+
+
+def build_key_value_pairs_parsed_output(**kwargs) -> Tuple[Dict[str, Any], str]:
+    """
+    A command might have multiple key value pairs, where the name of the key is an attribute of <KEY> element
+    this builder will create an output of those keys
+    Args:
+        **kwargs:
+                json_response (Dict): response received from Qualys API
+                command_data (Dict): data about a specific command from the general commands_data dict
+    Returns:
+        Tuple containing a dictionary created from those key-value pairs and a markdown table as a string
+    """
+    command_data = kwargs['command_data']
+    output = kwargs['json_response']
+
+    parsed_output = parse_key_value_pairs_list(output)
+    readable_output = tableToMarkdown(name=command_data['table_name'], t=parsed_output)
+
+    return parsed_output, readable_output
+
+
+''' COMMAND FUNCTIONS '''
+
+
+def test_module(client: Client) -> str:
+    """
+    This function makes a http request to qualys API in order to test the connection
+    Args:
+        client: Client object for making a http request
+    Returns:
+        'ok' message if the connection test was successful
+    Raises:
+        DemistoException: will be raised when connection was not successful by command_http_request
+    """
+    commands_data['args_values'] = {}
+    client.command_http_request(commands_data['test-module'])
+    return 'ok'
+
+
+def qualys_command(client: Client,
+                   args: Dict[str, str],
+                   command_name: str,
+                   command_methods: Dict[str, Callable]) -> Optional[CommandResults]:
+    """
+    A general function to run any qualys command
+    Args:
+        client: Client object for making a http request
+        args: Dictionary of the arguments entered by the user for the command
+        command_name: string of the command name
+        command_methods: Dictionary of handler and output builder of the specific command
+    Returns:
+        Results received by the command or None if it's a file download
+    Raises:
+        DemistoException: will be raised when request to Qualys API failed
+    """
+    build_args_dict(args, commands_data[command_name], False)
+    build_args_dict(args, commands_data[command_name], True)
+    result = client.command_http_request(commands_data[command_name])
+    return command_methods['result_handler'](result=result,
+                                             command_name=command_name,
+                                             output_builder=command_methods.get('output_builder'))
+
+
+''' MAIN FUNCTION '''
+
+
+def main():
+    params = demisto.params()
+
+    base_url = params['url']
+    verify_certificate = not params.get('insecure', False)
+    proxy = params.get('proxy', False)
+    username = params.get('credentials').get('identifier')
+    password = params.get('credentials').get('password')
```

Comment:
I first used that because it was generated automatically when I created the integration, so I thought that it's a standard used by us, it also seems that it is used by other integrations as well.

I agree that the safe get is not necessary, however with the current situation the user will make a request to Qualys and will receive an error for missing credentials. So should I remove this and handle the error for missing credentials myself?

Reply:
Aren't those parameters required? Is there any situation that they will not exist?

Reply:
I changed it to direct access and then I got a warning while creating my build

> W9019 main: It is best practice to use .get when accessing the arg/params dict object rather then direct access.

So i'll guess I use the safe get